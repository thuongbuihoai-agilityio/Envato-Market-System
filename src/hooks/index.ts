export * from './useForm';
export * from './useAuth';
export * from './useStatistical';
export * from './useTransactions';
export * from './useEmployee';
export * from './useDebounce';
export * from './useSearch';
export * from './usePagination';
export * from './useUser';
