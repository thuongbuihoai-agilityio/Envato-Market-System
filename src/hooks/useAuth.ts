import { useCallback } from 'react';
import { AxiosResponse } from 'axios';
import { shallow } from 'zustand/shallow';

// Constants
import {
  END_POINTS,
  SEARCH_PARAM,
  ERROR_MESSAGES,
  IMAGES,
} from '@app/constants';

// Services
import { UsersHttpService } from '@app/services';

// Types
import { TUserDetail } from '@app/interfaces/user';

// Utils
import { getCurrentTimeSeconds } from '@app/utils';

// Stores
import { authStore } from '@app/stores';

type TSignUpErrorField = Partial<
  Record<keyof Omit<TUserDetail, 'id' | 'createdAt'>, string>
>;

export type TUserInfo = Omit<TUserDetail, 'password'> | null;

export type TUseAuth = {
  user: TUserInfo;
  isRemember: boolean;
  date: number;
  setUser: (user: TUserDetail) => void;
  signIn: (
    {
      email,
      password,
    }: {
      email: string;
      password: string;
    },
    isRemember?: boolean,
  ) => Promise<void>;
  signUp: (userInfo: Omit<TUserDetail, 'id' | 'createdAt'>) => Promise<{
    errors?: TSignUpErrorField;
  }>;
  signOut: () => void;
};

export const useAuth = () => {
  const { updateStore, clearStore } = authStore(
    (state) => ({
      updateStore: state.updateStore,
      clearStore: state.clearStore,
    }),
    shallow,
  );

  const handleSignIn = useCallback(
    async (
      {
        email,
        password,
      }: {
        email: string;
        password: string;
      },
      isRemember?: boolean,
    ): Promise<void> => {
      const { data = [] }: AxiosResponse<TUserDetail[] | undefined> =
        await UsersHttpService.get<TUserDetail[] | undefined>(
          `${END_POINTS.USERS}?${SEARCH_PARAM.EMAIL}=${email}&${SEARCH_PARAM.PASSWORD}=${password}`,
        );

      // Because search by params working incorrect
      const user: TUserDetail | undefined = data.find(
        (user) => user.email === email && user.password === password,
      );

      if (!user) {
        throw new Error(ERROR_MESSAGES.AUTH_INCORRECT);
      }

      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { password: pass, ...userInfo } = user;

      return updateStore({
        user: userInfo,
        isRemember,
        date: getCurrentTimeSeconds(),
      });
    },
    [updateStore],
  );

  const handleSignUp = useCallback(
    async (
      userInfo: Omit<TUserDetail, 'id' | 'createdAt'>,
    ): Promise<{
      errors?: TSignUpErrorField;
    }> => {
      const { email, password } = userInfo;
      const { data = [] }: AxiosResponse<TUserDetail[] | undefined> =
        await UsersHttpService.get<TUserDetail[] | undefined>(
          `${END_POINTS.USERS}?${SEARCH_PARAM.EMAIL}=${email}&${SEARCH_PARAM.PASSWORD}=${password}`,
        );

      // Because search by params working incorrect
      const user: TUserDetail | undefined = data.find(
        (user) => user.email === email,
      );

      if (user) {
        return {
          errors: {
            email: ERROR_MESSAGES.ACCOUNT_ALREADY_EXISTS,
          },
        };
      }

      // Send request add new user
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { password: responsePassword, ...response }: TUserDetail =
        await UsersHttpService.post<TUserDetail>(
          END_POINTS.USERS,
          {
            ...userInfo,
            avatarURL: IMAGES.AVATAR_SIGN_UP.url,
            createdAt: Date.now(),
          },
          {},
        ).then((res) => res.data);

      // Save user into store
      updateStore({ user: response, date: getCurrentTimeSeconds() });

      return {};
    },
    [updateStore],
  );

  return {
    setUser: updateStore,
    signIn: handleSignIn,
    signUp: handleSignUp,
    signOut: clearStore,
  };
};
