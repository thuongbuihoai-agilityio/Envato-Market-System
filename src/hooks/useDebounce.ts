import { MutableRefObject, useCallback, useRef, DependencyList } from 'react';

// Constants
import { DEBOUNCE_TIME } from '@app/constants';

export const useDebounce = <TParam = unknown>(
  callback: (...args: TParam[]) => void,
  dependencies: DependencyList,
) => {
  const refTime: MutableRefObject<NodeJS.Timeout | null> = useRef(null);

  return useCallback((...args: TParam[]) => {
    if (refTime.current) clearTimeout(refTime.current);

    refTime.current = setTimeout(() => {
      callback(...args);
    }, DEBOUNCE_TIME);
  }, dependencies);
};
