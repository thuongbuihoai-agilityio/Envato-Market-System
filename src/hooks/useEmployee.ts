import { useQuery } from '@tanstack/react-query';

// Services
import { getEmployees } from '@app/services';

export const useEmployee = () => {
  const query = useQuery({
    queryKey: [''],
    queryFn: getEmployees,
  });

  // TODO: update return value later
  return query;
};
