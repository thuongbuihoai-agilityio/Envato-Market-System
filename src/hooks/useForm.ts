import {
  FieldValues,
  UseFormProps,
  useForm as useReactHookForm,
} from 'react-hook-form';

export const useForm = <
  TFieldValues extends FieldValues = FieldValues,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  TContext = any,
  TTransformedValues extends FieldValues | undefined = undefined,
>(
  option?: UseFormProps<TFieldValues, TContext>,
) =>
  useReactHookForm<TFieldValues, TContext, TTransformedValues>({
    ...option,
    reValidateMode: 'onBlur',
  });
