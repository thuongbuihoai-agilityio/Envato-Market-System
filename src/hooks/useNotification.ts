import { Notification } from '@app/interfaces';
import { useCallback } from 'react';

export const useNotification = (notifications: Notification[]) => {
  const { quantity, hasNewNotification } = notifications.reduce(
    (result, notification) => {
      if (!notification?.isMarkRead) {
        result.quantity += 1;
        result.hasNewNotification = true;
      }
      return result;
    },
    { quantity: 0, hasNewNotification: false },
  );

  const handleDeleteNotice = useCallback(
    (id: string) => notifications.filter((item) => item.id !== id),

    [notifications],
  );

  const handleUpdateMarkRead = (id: string) => {
    const notificationToUpdate = notifications.find(
      (notification) => notification.id === id,
    );
    if (notificationToUpdate) {
      notificationToUpdate.isMarkRead = true;
    }
  };

  return {
    quantity,
    hasNewNotification,
    handleDeleteNotice,
    handleUpdateMarkRead,
  };
};
