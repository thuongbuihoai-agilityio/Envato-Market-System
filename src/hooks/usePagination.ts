import { useCallback, useMemo, useState } from 'react';

// Types
import { TTransaction } from '@app/interfaces';

export const usePagination = (transactions: TTransaction[]) => {
  const [data, setData] = useState({
    limit: 10,
    currentPage: 1,
  });

  const { limit, currentPage } = data;

  const filterData = useMemo(() => {
    const start = (currentPage - 1) * limit;
    const end = limit + start;

    return transactions.slice(start, end);
  }, [currentPage, limit, transactions]);

  const handleChangeLimit = useCallback(
    (limit: number) => {
      setData({
        ...data,
        currentPage: (data.currentPage = 1),
        limit,
      });
    },
    [data],
  );

  const handleChangePage = useCallback(
    (currentPage: number) => {
      setData({
        ...data,
        currentPage,
      });
    },
    [data],
  );

  const handleSearchWithPagination = useCallback(
    (searchTransactionValue = '', onSearchTransaction: () => void) => {
      setData((prevData) => ({
        ...prevData,
        currentPage:
          transactions.length < data.limit || !searchTransactionValue
            ? 1
            : prevData.currentPage,
      }));
      onSearchTransaction();
    },
    [transactions.length, data.limit, setData],
  );

  const resetPage = useCallback(
    () => setData((prev) => ({ ...prev, currentPage: 1 })),
    [],
  );

  return {
    data,
    filterData,
    setData,
    resetPage,
    handleChangeLimit,
    handleChangePage,
    handleSearchWithPagination,
  };
};
