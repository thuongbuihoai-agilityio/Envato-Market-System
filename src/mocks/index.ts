export * from './spending';
export * from './select';
export * from './table';
export * from './sidebar';
