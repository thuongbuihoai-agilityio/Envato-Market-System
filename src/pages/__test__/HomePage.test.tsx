import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import DashboardPage from '@app/pages/Home/index';
import { INITIAL_EFFICIENCY } from '@app/mocks';
import { useGetStatistic } from '@app/hooks';

// Mock the modules or functions used in your component
jest.mock('@app/hooks', () => ({
  useGetStatistic: jest.fn(),
}));

describe('Dashboard Component', () => {
  (useGetStatistic as jest.Mock).mockReturnValue({
    data: INITIAL_EFFICIENCY,
    isLoading: false,
    isError: false,
  });

  it('Should render match with snapshot.', async () => {
    const { container } = render(<DashboardPage />);

    expect(container).toMatchSnapshot();
  });
});
