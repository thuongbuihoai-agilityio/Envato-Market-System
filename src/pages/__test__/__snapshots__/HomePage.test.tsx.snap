// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component Should render match with snapshot. 1`] = `<div />`;
